src/hotspot/share/gc/serial/defNewGeneration.cpp

```diff
@@ -669,9 +669,21 @@ void DefNewGeneration::init_assuming_no_promotion_failure() {
 }
 
 void DefNewGeneration::remove_forwarding_pointers() {
-  RemoveForwardedPointerClosure rspc;
-  eden()->object_iterate(&rspc);
-  from()->object_iterate(&rspc);
+  assert(_promotion_failed, "precondition");
+
+  // Will enter Full GC soon due to failed promotion. Must reset the mark word
+  // of objs in young-gen so that no objs are marked (forwarded) when Full GC
+  // starts. (The mark word is overloaded: `is_marked()` == `is_forwarded()`.)
+  struct ResetForwardedMarkWord : ObjectClosure {
+    void do_object(oop obj) override {
+      if (obj->is_forwarded()) {
+        obj->init_mark();
```

Comment:
Since it is available, probably use `PreservedMarks::init_forwarded_mark` here.

Reply:
This closure is not really for "preserved marks", so calling `PreservedMarks::XXX` can be confusing. I will keep the current code and investigate the possibility of removing `PreservedMarks::init_forwarded_mark` in another PR --- its caller should not know anything about "preserved marks" either.